docs/tutorials/hpo-workflows/hpo-workflows.rst

```diff
@@ -0,0 +1,334 @@
+
+                        self.log(np.mean(train_losses), "train_loss", kind="metric", step=epoch)
+                        self.log(np.mean(val_losses), "val_loss", kind="metric", step=epoch)
+                        checkpoint = {
+                            "epoch": epoch,
+                            "loss": train_loss,
+                            "val_loss": val_loss,
+                        }
+                        ############### This is unique to the RayTorchTrainer ##################
+                        metrics = {"loss": val_loss.item()}
+                        self.checkpoint_and_report(
+                            epoch, tuning_metrics=metrics, checkpointing_data=checkpoint
+                        )
+                        ########################################################################
+
+
+    .. tab:: TorchTrainer
+
+        .. code-block:: python
+
+            class MyTrainer(TorchTrainer):
+                def __init__(
+                    self,
+                    config: Dict | TrainingConfiguration | None = None,
+                    strategy: Literal["ddp", "deepspeed", "horovod"] = "ddp",
+                    name: str | None = None,
+                    logger: Logger | None = None,
+                ) -> None:
+                    self.config = config
+                    super().__init__(config=config, strategy=strategy, name=name, logger=logger)
+
+                def create_model_loss_optimizer(self):
+                    model = resnet18(num_classes=10)
+                    model.conv1 = torch.nn.Conv2d(
+                        1, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False
+                    )
+                    # First, define strategy-wise optional configurations
+                    if isinstance(self.strategy, DeepSpeedStrategy):
+                        distribute_kwargs = dict(
+                            config_params=dict(train_micro_batch_size_per_gpu=self.config.batch_size)
+                        )
+                    else:
+                        distribute_kwargs = {}
+                    optimizer = Adam(model.parameters(), lr=self.config.learning_rate)
+                    self.model, self.optimizer, _ = self.strategy.distributed(
+                        model, optimizer, **distribute_kwargs
+                    )
+                    self.loss = CrossEntropyLoss()
+
+                def train(self, config, data):
+                    
+                    for epoch in range(self.config.epochs):
+                        if self.strategy.global_world_size() > 1:
+                            self.set_epoch(epoch)
+
+                        train_losses = []
+                        val_losses = []
+
+                        for images, labels in enumerate(self.train_dataloader):
+                            
+                            device = self.strategy.device()
+                            images, labels = images.to(device), labels.to(device)
```

Comment:
Ref: above conversation: 

It seems to me that you do move them to device with the regular trainer, so maybe change that in the RayTorchTrainer so as to not confuse anyone?

Reply:
See above :)